server/src/main/java/org/elasticsearch/action/admin/indices/analyze/TransportAnalyzeAction.java

```diff
@@ -163,7 +164,9 @@ private static Analyzer getAnalyzer(AnalyzeAction.Request request, AnalysisRegis
                 if (analyzer == null) {
                     throw new IllegalArgumentException("failed to find global analyzer [" + request.analyzer() + "]");
                 }
-                return analyzer;
+                return analyzer instanceof NamedAnalyzer
+                       ? new NamedAnalyzer((NamedAnalyzer)analyzer, TextFieldMapper.Defaults.POSITION_INCREMENT_GAP)
+                       : analyzer;
```

Comment:
This fix seems to work, but I wonder if it's a bit fragile. What is the underlying difference between the case with an index and that without an index? Is there a way to unify the too somehow?

Reply:
Thanks @javanna for the feedback. 

When reaching the  **TransportAnalyzeAction** class, the method **getAnalyzer** retrieves the target Analyzer. When the **index** is provided, the **indexService** is called, which has been bootstrapped during the **AnalysisRegistry::build::produceAnalyzer**, overriding the gap to **100**.

Without an **index** in the rest call, the **AnalysisRegistry** returns the analyzer. In that case, the returned object can be of type **NamedAnalyzer** (the gap is overridden to Integer.MIN_VALUE) or a plain **Lucene** Analyzer object (gap defaults to 0). For both cases, no method is available to override the value of the gap.

I identified three ways to solve it so far.

- The first was to override the NameAnalyzer after retrieving it from the AnalysisRegistry. This applies only to the objects of type NamedAnalyzer
- The second could be to hardcode the incrementGap into the calculation method TransportAnalyzeAction::simpleAnalyze to add 100 when detects, e.g., a negative number
- Another way would be to create NamedAnalyzer objects with the value 100 instead of Inger:Min. 

I haven't yet considered unifying the logic. Further investigation is required for that.

Reply:
Sorry for the drive-by commenting, just because @drempapis and I talked about this earlier:

> The second could be to hardcode the incrementGap into the calculation method

while that's possible I noted that such an approach would need to take care of both "simpleAnalyze()" and detailAnalyze()" branches, so two places that would need changing and consideration.
I would say the wrapping approach looks the most reasonable to me, looking at the code I only wondered when we'd ever be in a position there the non-index code branch would return a NamedAnalyzer from the registry? Do you have a specific case for that at hand? If that's not possible (and I think we only create NamedAnalyzers as part of index analyzers at the moment), then an even more light-weight version instead of wrapping with a NamedAnalyzer would be to wrap in a DelegatingAnalyzerWrapper and only overwrite "getPositionIncrementGap" with the ES default of 100. As far as I can tell the only way to modify this value is in text field mappings which in turn should only be there if we specify an index.

Reply:
Thanks for the explanation. How difficult would it be to fake an IndexService when an index is not provided instead?

Reply:
Creating a fake IndexService is complex and requires many objects to mock.

https://github.com/elastic/elasticsearch/blob/8afe2f4f58a8fe709a192b69011dc447c7454658/server/src/main/java/org/elasticsearch/index/IndexService.java#L167

Internally overrides the  **NamedAnalyzer** object similarly to how we do it here. Is there any other way of creating an IndexService that I may have missed?

https://github.com/elastic/elasticsearch/blob/8afe2f4f58a8fe709a192b69011dc447c7454658/server/src/main/java/org/elasticsearch/index/analysis/AnalysisRegistry.java#L723

I experimented with the following approach of creating a **dummyIndexService** in the **TransportAnalyzerAction** class.

```
    private static IndexService craeteDummyIndexService(IndicesService indicesService) throws IOException {
        Index index = new Index(UUIDs.randomBase64UUID(), UUIDs.randomBase64UUID());
        Settings.Builder builder = Settings.builder()
            .put(IndexMetadata.SETTING_VERSION_CREATED, IndexVersion.current())
            .put(SETTING_INDEX_UUID, index.getUUID());

        IndexMetadata indexMetadata = new IndexMetadata.Builder(index.getName()).settings(builder.build())
            .numberOfShards(1)
            .numberOfReplicas(0)
            .build();
         return indicesService.createIndex(indexMetadata, Collections.emptyList(), false);
    }
```

This approach can return correctly for the non-index _analyzer api-calls but fails eventually with the exception.
```
[2024-11-05T10:45:21,605][ERROR][o.e.b.ElasticsearchUncaughtExceptionHandler] [runTask-0] fatal error in thread [elasticsearch[runTask-0][clusterApplierService#updateTask][T#1]], exiting java.lang.AssertionError: index [ClMZwlqSQKSIeOZd-cgguA/D8sWGYUfSQu5vd9HQsVgvA] does not exist in the cluster state, it should either have been deleted or the cluster must be new
        at org.elasticsearch.server@9.0.0-SNAPSHOT/org.elasticsearch.indices.cluster.IndicesClusterStateService.removeIndicesAndShards(IndicesClusterStateService.java:468)
```

Reply:
> Sorry for the drive-by commenting, just because @drempapis and I talked about this earlier:
> 
> > The second could be to hardcode the incrementGap into the calculation method
> 
> while that's possible I noted that such an approach would need to take care of both "simpleAnalyze()" and detailAnalyze()" branches, so two places that would need changing and consideration. I would say the wrapping approach looks the most reasonable to me, looking at the code I only wondered when we'd ever be in a position there the non-index code branch would return a NamedAnalyzer from the registry? Do you have a specific case for that at hand? If that's not possible (and I think we only create NamedAnalyzers as part of index analyzers at the moment), then an even more light-weight version instead of wrapping with a NamedAnalyzer would be to wrap in a DelegatingAnalyzerWrapper and only overwrite "getPositionIncrementGap" with the ES default of 100. As far as I can tell the only way to modify this value is in text field mappings which in turn should only be there if we specify an index.

Thank you, @cbuescher, for reviewing this. When the method **AnalysisRegistry::getAnalyzer** is called, the prebuiltAnalysis is already available, having set the positionIncrementGap=INTEGER_MIN.
I don't see how we can wrap in a DelegatingAnalyzerWrapper, this class is abstract and does not provide a method to override the GAP.
```
 public Analyzer getAnalyzer(String analyzer) throws IOException {
        AnalysisModule.AnalysisProvider<AnalyzerProvider<?>> analyzerProvider = this.prebuiltAnalysis.getAnalyzerProvider(analyzer);
        if (analyzerProvider == null) {
            AnalysisModule.AnalysisProvider<AnalyzerProvider<?>> provider = analyzers.get(analyzer);
            return provider == null ? null : cachedAnalyzer.computeIfAbsent(analyzer, (key) -> {
                try {
                    return provider.get(environment, key).get();
                } catch (IOException ex) {
                    throw new ElasticsearchException("failed to load analyzer for name " + key, ex);
                }
            });
        }
        return analyzerProvider.get(environment, analyzer).get();
    }
```

Reply:
Thanks for considering faking the index service. I agree that's not easy. I looked a bit deeper at AnalysisRegistry, and I see that it already has the logic for wrapping the NamedAnalyzer in produceAnalyzer. I wonder if we should leverage that somehow in the analyze api as well. I see that getAnalyzer is only used by the analyze api, and the categorize text agg. I even wonder if the latter may benefit from the same fix around position increment gap. So, should the fix be rather made in getAnalyzer, and perhaps the code shared between produceAnalyzer and getAnalyzer, to make sure they stay in sync?

Reply:
Thank you, @javanna. I added that logic to the API call defensively, but I'll also consider the option, verify how the categorize text agg works, and unify the logic.

Reply:
@javanna I changed the `getAnalyzer` method as proposed, which works well for both the Analyze API and Categorize Text Agg. When an `index` is bootstrapped, the `produceAnalyzer` is called, requiring many objects to build and pass in the function call. Using the produceAnalyzer method when calling the `getAnalyzer` (in the case of no index provided) would result in more significant overhead and many intermediate classes to construct, which is unnecessary for this case.

Reply:
I did not necessarily mean that we need to call produceAnalyzer, but we could share the wrapping part between the two methods?